proposals/client-interceptors.md

```diff
@@ -0,0 +1,312 @@
+
+Additionally, the `options` argument includes a `MethodDescriptor` object. The `MethodDescriptor` is a container for properties of the call which are used internally and may also be useful to interceptors:
+
+```javascript
+/**
+ * @param {string} name The name of the call, i.e. 'myCall' 
+ * @param {string} serviceName The name of the service, i.e. 'MyService'
+ * @param {string} path The full path of the call, i.e. '/MyService/MyCall'
+ * @param {MethodType} methodType One of two types: MethodType.UNARY or MethodType.SERVER_STREAMING
+ * @param {grpc.web.AbstractClientBase.MethodInfo} methodInfo Properties containing the serialize and deserialize functions
+ */
+MethodDescriptor(name, serviceName, path, methodType, methodInfo)
+```
+*Do not modify the `options.methodDescriptor` object. The MethodDescriptor passed to the interceptor options is not consumed by the underlying gRPC code and changes will only affect downstream interceptors*
+
+A `requester` object is a plain Javascript object implementing zero or more outbound interception methods:
+
+```javascript
+/**
+ * An interception method called before an outbound call has started.
+ * @param {grpc.web.Metadata} metadata The call's outbound metadata (request headers).
+ * @param {object} listener The listener which will be intercepting inbound operations.
+ * @param {Function} next A callback which continues the gRPC interceptor chain, called with the metadata to send and the inbound listener.
+ */
+ start(metadata, listener, next)
+```
+```javascript
+/**
+ * An interception method called prior to every outbound request.
+ * @param {Object} request protobuf message
+ * @param {Function} next A callback which continues the gRPC interceptor chain, called with the request to send.
+ */
+send(request, next)
+```
+
+A `listener` object implements zero or more methods for intercepting inbound operations. The listener passed into a requester's `start` method implements all the inbound interception methods. Inbound operations will be passed through a listener at each step in the interceptor chain. Three usage patterns are supported for listeners:
+
+1. Pass the listener along without modification: `next(metadata, listener)`. In this case the interceptor declines to intercept any inbound operations.
+1. Create a new listener with one or more inbound interception methods and pass it to next. In this case the interceptor will fire on the inbound operations implemented in the new listener.
+1. Store the listener to make direct inbound calls on later. This effectively short-circuits the interceptor stack. Short-circuiting a request in this way will skip interceptors which have not yet fired, but will fire the listeners on any 'earlier' interceptors in the stack.
+
+*Do not modify the listener passed in. Either pass it along unmodified or call methods on it to short-circuit the interceptor stack.*
+
+The `listener` methods are:
```

Comment:
So, I'm not 100% convinced on the `listener` API for inbound operations. 
This has been ported from the NodeJS API; however I believe that this is over-complicated for the web implementation.
The way I see it, there only a few places where we need to intercept:

https://github.com/grpc/grpc-web/blob/master/javascript/net/grpc/web/grpcwebclientbase.js#L66
```javascript
GrpcWebClientBase.prototype.rpcCall = function(
    method, request, metadata, methodInfo, callback) {
  // Intercept HERE onStart()
  // ...
  stream.on('data');   // Intercept HERE onData()
  stream.on('status'); // Intercept HERE onStatus()
  stream.on('error'); // Intercept HERE onError()
  // ...
  // Intercept HERE onSend()
  xhr.send(method, 'POST', payload);
  return stream;
};
```
The `serverStreaming` is similar.

This would simplify the `requester` object to:

```javascript
var requester = {
  onStart: function(metadata, next){},
  onSend: function(request, next){},
  onData: function(response, next){},
  onStatus: function(status, next){},
  onError: function(error, next){},
};
```

Reply:
A case can be made to keep the API design consistent with other platforms. That way developers don't need to re-learn concepts. For instance, I find it convenient that the Java API[1] is the same for Android and server. A ClientCall in Java roughly translates to InterceptingCall as defined in this proposal. So defining a Listener means the JS API looks similar to the Java API.

OTOH, the ObjC Interceptor API proposal does *not* define the concept of Listener and instead puts all methods on GRPCInterceptorInterface[2] instead.

[1]https://grpc.github.io/grpc-java/javadoc/io/grpc/ClientCall.html#start-io.grpc.ClientCall.Listener-io.grpc.Metadata-
[2]https://github.com/grpc/proposal/blob/9bcd06aae8e6f9122885711a602ecdc9af75c560/L50-objc-interceptor.md#interceptor-chain